[WIP] Jellyfin port

Deanosim opened the pull request:
#1 Partially working, all the menus work and you can enter your tunein username to see your personal stations, but media playback won't work on either web or android.

Deanosim committed:
port to jellyfin

Changed namespace and logging entries
Plugin is only partly functional, all the menus work but the media playback does not.

Deanosim committed:
Cleanup old files

cvium commented:
What's with all the changes to HtmlAgilityPack? You should also add `obj/` to .gitignore as it's not something you want to commit

cvium commented:
And you should pull in HtmlAgilityPack from nuget and get rid of the source in the repo

Deanosim committed:
remove htmlagilitypack source, add obj/ to gitignore

Removed HtmlAgilityPack source and replaced with nuget.
Added obj/ to .gitignore

Deanosim committed:
Update .gitignore

Deanosim committed:
Update .gitignore

Deanosim committed:
Fixed gitignore and removed obj folder

Deanosim commented:
Forgot to mention that I fixed all that, HtmlAgilityPack source is gone and uses nuget, and obj is added to the gitignore file.

cvium commented on Jellyfin.Plugins.TuneIn/Jellyfin.Plugins.TuneIn.csproj:
```diff
@@ -1,44 +1,45 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>netstandard2.0;</TargetFrameworks>
```
Should be
```suggestion
    <TargetFramework>netstandard2.0</TargetFramework>
```

cvium commented on Jellyfin.Plugins.TuneIn/TuneInChannel.cs:
```diff
@@ -1,529 +1,529 @@
+            _appHost = appHost;
+            _logger = loggerFactory.CreateLogger(GetType().Name);
+
+            partnerid = "uD1X52pA";
```
Do we need our own partner id?

Deanosim replied:
Did a bit of reading and I can't find much information for it, There's an API for broadcasters but that seems to be about it, so maybe we need to contact them.
https://tunein.com/broadcasters/api/

cvium replied:
Cc @anthonylavado

Deanosim replied:
A friend found this on archive.org 
http://web.archive.org/web/20110823171916/http://inside.radiotime.com/developers/api/opml

![](https://cdn.discordapp.com/attachments/307000293483347992/544155947497685014/unknown.png)

anthonylavado replied:
@Deanosim Good find! I’ve been struggling to locate anything at all for a while. I’ll contact them.

anthonylavado replied:
Just a heads up here - I have not heard back from that e-Mail address. It didn’t bounce back either... I’ll make an issue for tracking this.

WillWill56 replied:
Thanks, yeah that information is a bit old, I wonder if anyone at TuneIn is still monitoring that address... may need to try another contact method if you never hear back.

cvium commented on Jellyfin.Plugins.TuneIn/TuneInChannel.cs:
```diff
@@ -1,529 +1,529 @@
+        public String GetDefaultImages(String name)
+        {
+            if (name == "Local Radio")
+                return "https://raw.githubusercontent.com/snazy2000/MediaBrowser.Channels/master/MediaBrowser.Plugins.TuneIn/Images/tunein-localradio.png";
```
These should be changed

Deanosim replied:
Is their a Jellyfin fork for the Channels Emby github? I can't seem to find one.

cvium replied:
This one https://github.com/jellyfin/jellyfin-plugin-channels? It's not ready though. There's an ongoing pr

cvium commented:
I'll take a closer look once I'm behind a computer again

cvium committed:
Update Jellyfin.Plugins.TuneIn/Jellyfin.Plugins.TuneIn.csproj

Co-Authored-By: Deanosim <4834976+Deanosim@users.noreply.github.com>

WillWill56 committed:
Updated some urls

WillWill56 committed:
Add Visual Studio Code to .gitignore

WillWill56 committed:
Temporarily remove partnerid

Deanosim commented:
So weirdly enough the plugin seems to have stopped working since last night, the tunein menu item just displays a blank screen now, and my friend @WillWill56 who's been helping me test is also getting the same thing, so something has either changed or broken. But I can't figure out what.

hawken93 commented:
Git bisect can help you if you find a working commit :)

Deanosim commented:
Yep original port works so it was removing htmlagilitypack source code and turning it into a NuGet, I don't understand why but that's probably due to my limited knowledge on c# and .net

WillWill56 commented:
The main problem at the moment is satisfying the HtmlAgilityPack dependency on the server. Simply copying the HtmlAgilityPack.dll file into the plugins folder or into jellyfin's root directory doesn't seem to be enough, it may have its own dependencies that aren't being met (quite a few to choose from in the publish directory). Could potentially ditch HtmlAgilityPack and use XmlDocument instead.

cvium commented:
Take a look at jellyfin#864. The change from Assembly.Load to Assembly.LoadFrom allows dynamic dependency resolving

Deanosim committed:
Update Jellyfin.Plugins.TuneIn.csproj

cvium commented:
864 has just been merged into release-10.2.z so it should work now

WillWill56 committed:
Add some exception handling for debugging

WillWill56 committed:
Change to Jellyfin namespace

WillWill56 commented:
jellyfin#864 solved that problem as you said it would, only need the two dll files in a subdirectory and it works. Seem to have another problem with jellyfin though. In the file MediaBrowser.Api/Playback/BaseStreamingService.cs, there's an IHttpClient that gets used, but it seems like nothing is ever assigned to it, so is potentially causing the null reference exceptions I keep getting (log excerpt below).

```
[2019-02-11 11:18:48.885 +00:00] [INF] HttpClientManager "GET": "http://opml.radiotime.com/Tune.ashx?id=s7007&amp;formats=mp3,aac&amp;serial=59ac668738114b9196b2bea0c63a32a0&formats=mp3,aac"
[2019-02-11 11:18:51.385 +00:00] [INF] Normal URL
[2019-02-11 11:18:51.386 +00:00] [INF] TuneIn Found Direct Media Source
[2019-02-11 11:18:51.387 +00:00] [INF] Normal URL
[2019-02-11 11:18:51.388 +00:00] [INF] TuneIn Found Direct Media Source
[2019-02-11 11:18:51.388 +00:00] [INF] TuneIn Found Media source: http://shout01.stream.hostworks.com.au/3aw
[2019-02-11 11:18:51.389 +00:00] [INF] TuneIn Found Media source: http://shout01.stream.hostworks.com.au/3aw_mp3
[2019-02-11 11:18:51.391 +00:00] [INF] User policy for "William". EnableAudioPlaybackTranscoding: True
[2019-02-11 11:18:51.392 +00:00] [INF] User policy for "William". EnableAudioPlaybackTranscoding: True
[2019-02-11 11:18:51.396 +00:00] [ERR] Error processing request
System.NullReferenceException: Object reference not set to an instance of an object.
   at MediaBrowser.Api.Playback.Progressive.BaseProgressiveStreamingService.GetStaticRemoteStreamResult(StreamState state, Dictionary`2 responseHeaders, Boolean isHeadRequest, CancellationTokenSource cancellationTokenSource) in /repo/MediaBrowser.Api/Playback/Progressive/BaseProgressiveStreamingService.cs:line 303
   at MediaBrowser.Api.Playback.Progressive.BaseProgressiveStreamingService.ProcessRequest(StreamRequest request, Boolean isHeadRequest) in /repo/MediaBrowser.Api/Playback/Progressive/BaseProgressiveStreamingService.cs:line 170
   at MediaBrowser.Api.Playback.UniversalAudioService.GetUniversalStream(GetUniversalAudioStream request, Boolean isHeadRequest) in /repo/MediaBrowser.Api/Playback/UniversalAudioService.cs:line 362
   at Emby.Server.Implementations.Services.ServiceExecGeneral.GetTaskResult(Task task) in /repo/Emby.Server.Implementations/Services/ServiceExec.cs:line 110
   at Emby.Server.Implementations.Services.ServiceHandler.ProcessRequestAsync(HttpListenerHost appHost, IRequest httpReq, IResponse httpRes, ILogger logger, String operationName, CancellationToken cancellationToken) in /repo/Emby.Server.Implementations/Services/ServiceHandler.cs:line 95
   at Emby.Server.Implementations.HttpServer.HttpListenerHost.RequestHandler(IHttpRequest httpReq, String urlString, String host, String localPath, CancellationToken cancellationToken) in /repo/Emby.Server.Implementations/HttpServer/HttpListenerHost.cs:line 608
```

WillWill56 committed:
Silently ignore duplicate keys in .pls file

WillWill56 committed:
Fix case sensitive urls in .pls files

WillWill56 committed:
Fix GetChannelItemMediaInfo request url

WillWill56 committed:
minor rewrite of GetMenu

Deanosim committed:
Merge pull request #1 from WillWill56/jellyfin_port

Some bugfixes and small improvements

Deanosim commented:
Thanks to @WillWill56 for his work, This is kinda out of my depth at this point :P

Deanosim committed:
Revert to Jellyfin 10.1 so builds work

WillWill56 committed:
fix namespace

WillWill56 committed:
fix loop, prevent null path in media source info

WillWill56 committed:
temporary fix for some stations not playing

WillWill56 committed:
ignore non-media lines in m3u files

WillWill56 commented:
Speaking of issues, I'm guessing I should create an issue associated with each significant bug-fix or improvement I've made in porting this plugin to Jellyfin? Can do that next time I get a chance to work on this.

WillWill56 committed:
additional handling for playlist files

WillWill56 committed:
further simplify GetMenu

Deanosim committed:
Merge pull request #2 from WillWill56/jellyfin_port

More fixes

WillWill56 committed:
handle lone Stations subcategory

WillWill56 committed:
update readme, build.yaml, version, guid...

WillWill56 committed:
bump controller version to 10.2.0

WillWill56 committed:
use embedded images

Deanosim committed:
Merge pull request #3 from WillWill56/jellyfin_port

Final changes & housekeeping
